parsl/data_provider/staging.py

```diff
@@ -0,0 +1,68 @@
+from concurrent.futures import Future
+from typing import Optional, Callable
+from parsl.app.futures import DataFuture
+from parsl.data_provider.files import File
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from parsl.data_provider.data_manager import DataManager
+
+
+class Staging:
+    """
+    This class defines the interface for file staging providers.
+
+    For each file to be staged in, the data manager will present the file
+    to each configured Staging provider in turn: first, it will ask if the
+    provider can stage this file by calling `can_stage_in`, and if so, it
+    will call both `stage_in` and `replace_task` to give the provider the
+    opportunity to perform staging.
+
+    For each file to be staged out, the data manager will follow the same
+    pattern using the corresponding stage out methods of this class.
+
+    The default implementation of this class rejects all files, and
+    performs no staging actions.
+
+    To implement a concrete provider, one or both of the `can_stage_*`
```

Comment:
Should we enforce the interface with ABCs?

Reply:
There isn't anything to enforce here, I think -- it's fine to leave any of the methods non-overridden because they have "sensible" do-nothing defaults, and it is not likely that a particular provider would want to override all of the methods.

Reply:
:+1: